packages/ai/src/LLMStep.ts

```diff
@@ -186,26 +205,20 @@ export class LLMStepInstance<const Shape extends StepShape = StepShape> {
     this.logger.log(log);
   }
 
-  private criticalError(error: string) {
-    this.log({ type: "error", message: error });
+  private fail(errorType: ErrorType, message: string) {
+    this.log({ type: "error", message });
     this.state = {
       kind: "FAILED",
       durationMs: this.calculateDuration(),
-      error,
+      errorType,
+      message,
```

Comment:
There's some awkwardness here that makes me want to suggest exceptions as a failure mechanism.

I remember you don't like exceptions, so I'll explain my observations first.

1. You log the error here (reasonable), but then sometimes you log it again, e.g. in `FixCodeUntilItRuns`:
```typescript
        context.fail("MINOR", newCode.value);
        context.log({
          type: "codeRunError",
          error: newCode.value,
        });
```

This will lead to duplicate error log entries.

Because you needed to type it as `codeRunError`... ok, one way around it is to pass the error to `fail()`, like, `fail(errorType: ErrorType, message: Extract<LogEntry, { type: "error" | "codeRunError" }>)`, and then pass that. Fixable.

2. But then there's also the awkwardness that the step code can call `fail()` and then continue to do things. Which would be mostly ignored: the step is already failed, there should be nothing to do.

This also interferes with `LLMStep.run()`: you had to explicitly cast `this.state as CodeState` there, etc.

So, how about we just use `throw MinorStepError(message)` as the main way to fail? Then `run()` becomes more natural, if we catch an error then we minor-fail and either log the message or log `String(error)` if it's not a minor error. And we `complete()` otherwise.

One reason why I like this is because we _already_ support exceptions; steps can throw, and we don't ask step implementations to carefully try/catch their code. So they might as well throw more meaningful errors.

In other words, `fail("MINOR", "message")` function is unnecessary, it's identical to `throw new Error("message")`.

Reply:
Good idea, will experiment here.